src/data-workspace/use-data-value-mutation.js

```diff
@@ -0,0 +1,95 @@
+import { useDataEngine } from '@dhis2/app-runtime'
+import { useQueryClient, useMutation } from 'react-query'
+import { useContextSelection } from '../context-selection/index.js'
+import { dataValueQuery, useAttributeOptionCombo } from './data-workspace.js'
+
+const DATA_VALUE_MUTATION = {
+    resource: 'dataValues',
+    type: 'create',
+    params: ({ ...params }) => ({ ...params }),
+}
+
+export const useDataValueMutation = () => {
+    const queryClient = useQueryClient()
+    const [{ dataSetId, orgUnitId, periodId }] = useContextSelection()
+    const attributeOptionComboId = useAttributeOptionCombo()
+    const engine = useDataEngine()
+
+    const dataValueQueryKey = [
+        dataValueQuery,
+        {
+            dataSetId,
+            periodId,
+            orgUnitId,
+            attributeOptionComboId,
+        },
+    ]
+    const mutationFn = (variables) =>
+        engine.mutate(DATA_VALUE_MUTATION, { variables })
+
+    return useMutation(mutationFn, {
+        // Optimistic update of the react-query cache
+        onMutate: async (newDataValue) => {
+            // Cancel any outgoing refetches (so they don't overwrite our optimistic update)
+            await queryClient.cancelQueries(dataValueQueryKey)
+
+            // Snapshot the previous value
+            const previousDataValues =
+                queryClient.getQueryData(dataValueQueryKey)
+
+            // Optimistically update to the new value
+            queryClient.setQueryData(dataValueQueryKey, (oldDataValues) => {
+                // Ensure we don't mutate the snapshot
+                const newDataValues = {
+                    ...oldDataValues,
+                    dataValues: {
+                        ...oldDataValues.dataValues,
+                        dataValues: [...oldDataValues.dataValues.dataValues],
+                    },
+                }
+
+                newDataValues.dataValues.dataValues =
+                    newDataValues.dataValues.dataValues.map((dataValue) => {
+                        const {
+                            categoryOptionCombo,
+                            dataElement,
+                            orgUnit,
+                            period,
+                        } = dataValue
+                        const { co, de, ou, pe, value } = newDataValue
+                        const match =
+                            categoryOptionCombo === co &&
+                            dataElement === de &&
+                            orgUnit === ou &&
+                            period === pe
+
+                        if (!match) {
+                            return dataValue
+                        }
+
+                        return {
+                            ...dataValue,
+                            value,
+                        }
+                    })
+
+                return newDataValues
+            })
```

Comment:
```suggestion
            queryClient.setQueryData(dataValueQueryKey, (prevSnapshot) => {
                const updatedDataValues =
                    prevSnapshot.dataValues.dataValues.map((dataValue) => {
                        const {
                            categoryOptionCombo,
                            dataElement,
                            orgUnit,
                            period,
                        } = dataValue
                        const { co, de, ou, pe, value } = newDataValue
                        const match =
                            categoryOptionCombo === co &&
                            dataElement === de &&
                            orgUnit === ou &&
                            period === pe

                        if (!match) {
                            return dataValue
                        }

                        return {
                            ...dataValue,
                            value,
                        }
                    })

                return {
                    ...prevSnapshot,
                    dataValues: {
                        ...oldDataValues.dataValues,
                        dataValues: updatedDataValues,
                    },
                }
            })
```

Maybe this simplifies it a bit and reduces the possibility of accidentally overwriting the old snapshot?

Reply:
Adapted this a little, see 9f1c5bc321289d6e59f316de76d8c39221d3e7d4